python/src/space/core/ops/change_data.py

```diff
@@ -52,24 +52,26 @@ def read_change_data(
         f"End snapshot ID {end_snapshot_id} should not be lower than start "
         f"snapshot ID {start_snapshot_id}")
 
-  all_snapshot_ids = sorted(storage.snapshot_ids)
+  all_snapshot_ids = []
+  current_snapshot = storage.snapshot(end_snapshot_id)
+  while current_snapshot.snapshot_id >= start_snapshot_id:
+    all_snapshot_ids.insert(0, current_snapshot.snapshot_id)
+    if not current_snapshot.HasField("parent_snapshot_id"):
+      break
+    current_snapshot = storage.snapshot(current_snapshot.parent_snapshot_id)
+
   all_snapshot_ids_set = set(all_snapshot_ids)
 
   if start_snapshot_id not in all_snapshot_ids_set:
```

Comment:
Can we check start_snapshot_id == all_snapshot_ids[0] ? Then we don't need to create a set all_snapshot_ids_set